RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromZSRawProducer.cc

```diff
@@ -0,0 +1,300 @@
+
+    void fill(StripClusterizerAlgorithm::Det const& det, StripClusterizerAlgorithm::output_t::FastFiller& record);
+
+  private:
+    std::unique_ptr<sistrip::FEDBuffer> buffers[1024];
+
+    const FEDRawDataCollection& rawColl;
+
+    ClustersFromZSRawAlgorithm& clusterizer;
+
+    // March 2012: add flag for disabling APVe check in configuration
+    bool doAPVEmulatorCheck;
+
+// #define VISTAT
+#ifdef VISTAT
+    struct Stat {
+      Stat() : totDet(0), detReady(0), detSet(0), detAct(0), detNoZ(0), detAbrt(0), totClus(0) {}
+      int totDet;    // all dets
+      int detReady;  // dets "updated"
+      int detSet;    // det actually set not empty
+      int detAct;    // det actually set with content
+      int detNoZ;    // det actually set with content
+      int detAbrt;   // det aborted
+      int totClus;   // total number of clusters
+    };
+
+    mutable Stat stat;
+    void incTot(int n) const { stat.totDet = n; }
+    void incReady() const { stat.detReady++; }
+    void incSet() const { stat.detSet++; }
+    void incAct() const { stat.detAct++; }
+    void incNoZ() const { stat.detNoZ++; }
+    void incAbrt() const { stat.detAbrt++; }
+    void incClus(int n) const { stat.totClus += n; }
+    void printStat() const {
+      std::cout << "VI clusters " << stat.totDet << ',' << stat.detReady << ',' << stat.detSet << ',' << stat.detAct
+                << ',' << stat.detNoZ << ',' << stat.detAbrt << ',' << stat.totClus << std::endl;
+    }
+
+#else
+    static void zeroStat() {}
+    static void incTot(int) {}
+    static void incReady() {}
+    static void incSet() {}
+    static void incAct() {}
+    static void incNoZ() {}
+    static void incAbrt() {}
+    static void incClus(int) {}
+    static void printStat() {}
+#endif
+  };
+
+  class SiStripClusterizerFromZSRaw final : public edm::stream::EDProducer<> {
+  public:
+    explicit SiStripClusterizerFromZSRaw(const edm::ParameterSet& conf)
+        : productToken_(consumes<FEDRawDataCollection>(conf.getParameter<edm::InputTag>("ProductLabel"))),
+          doAPVEmulatorCheck_(conf.getParameter<bool>("DoAPVEmulatorCheck")) {
+      produces<edmNew::DetSetVector<SiStripCluster>>();
+    }
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+      edm::ParameterSetDescription desc;
+
+      desc.add<edm::InputTag>("ProductLabel", edm::InputTag("rawDataCollector"));
+      desc.add<bool>("DoAPVEmulatorCheck", false);
+
+      descriptions.add("SiStripClustersFromZSRawFacility", desc);
+    }
+
+    void beginRun(const edm::Run&, const edm::EventSetup& es) override { initialize(es); }
+
+    void produce(edm::Event& ev, const edm::EventSetup& es) override {
+      initialize(es);
+
+      // get raw data
+      edm::Handle<FEDRawDataCollection> rawData;
+      ev.getByToken(productToken_, rawData);
+
+      auto output = std::make_unique<edmNew::DetSetVector<SiStripCluster>>();
+      assert(!output->onDemand());
```

Comment:
isn't this superfluous given the line above?

Reply:
all assert are by definition superfluous